src/allocator/slice.rs

```diff
@@ -0,0 +1,145 @@
+//! Defines [`Slice`] --- list of buffers bounded by custome byte range.
+
+/// Represents bounded by custome range list of buffers.
+pub struct Slice {
+    buffers: Vec<Box<[u8]>>,
+    range: std::ops::Range<usize>,
+    sender: super::Sender<Box<[u8]>>,
+}
+
+impl Slice {
+    /// Returns new [`Slice`] of specified buffers.
+    ///
+    /// # Parameters
+    ///
+    /// - `buffers` --- vec of buffers.
+    /// - `range` --- range to which slice will allow access.
+    ///
+    /// # Panics
+    ///
+    /// This function will panics if called if length range bound greater then length of `buffers`.
+    pub fn new(
+        buffers: Vec<Box<[u8]>>,
+        range: std::ops::Range<usize>,
+        sender: super::Sender<Box<[u8]>>,
+    ) -> Self {
+        let len = buffers.iter().map(|buffer| buffer.len()).sum();
+
+        assert!(range.start <= len, "cannot index list as slice from start");
+        assert!(range.end <= len, "cannot index list as slice to end");
+
+        Self { buffers, range, sender }
+    }
+
+    pub fn iter_mut(&mut self) -> IterMut<'_> {
+        self.into_iter()
+    }
+
+    pub fn iter(&self) -> Iter<'_> {
+        self.into_iter()
+    }
+
+    /// Deallocates all buffers i.e. send them via specified in the [`Self::new`] `sender`.
+    fn deallocate(&mut self) {
+        for mut buffer in self.buffers.drain(..) {
+            // No user data should exist after dealloc
+            buffer.iter_mut().for_each(|u| *u = 0);
+            // Ignore allocator drop
+            let _ = self.sender.send(buffer);
+        }
+    }
+}
+
+impl Drop for Slice {
+    fn drop(&mut self) {
+        self.deallocate();
+    }
+}
+
+/// Shared iterator over [`Slice`] buffers.
+///
+/// Return shared slices accordingly to [`Slice`] bounds.
+pub struct Iter<'a> {
+    slice_iter: std::slice::Iter<'a, Box<[u8]>>,
+    range: std::ops::Range<usize>,
+}
+
+impl<'a> Iterator for Iter<'a> {
+    type Item = &'a [u8];
+
+    fn next(&mut self) -> Option<Self::Item> {
+        loop {
+            let result = self.slice_iter.next()?;
+
+            // save values for current step
+            let len = result.len();
+            let start = self.range.start;
+            let end = self.range.end;
+
+            // make a progress
+            self.range.start = self.range.start.saturating_sub(len);
+            self.range.end = self.range.end.saturating_sub(len);
+
+            if len > start {
+                return Some(&result[self.range.start..end.min(len)]);
```

Comment:
Shouldn't we use saved value?
```suggestion
                return Some(&result[start..end.min(len)]);
```